truffle/src/main/ruby/core/type.rb

```diff
@@ -210,6 +220,80 @@ def self.execute_check_convert_type(obj, cls, meth)
       raise TypeError, msg
     end
 
+    def self.rb_check_convert_type(obj, cls, meth)
+      return obj if object_kind_of?(obj, cls)
+      v = convert_type(obj, cls, meth, false)
+      return nil if v.nil?
+      unless object_kind_of?(v, cls)
+        raise TypeError, "Coercion error: obj.#{meth} did NOT return a #{cls} (was #{object_class(v)})"
+      end
+      v
+    end
+
+    def self.convert_type(obj, cls, meth, raise_on_error)
+      r = check_funcall(obj, meth)
+      if undefined.equal?(r)
+        if raise_on_error
+          raise TypeError, "can't convert #{obj} into #{cls} with #{meth}"
+        end
+        return nil
+      end
+      r
+    end
+
+    def self.check_funcall(recv, meth, args = [])
+      check_funcall_default(recv, meth, args, undefined)
+    end
+
+    def self.check_funcall_default(recv, meth, args, default)
+      respond = check_funcall_respond_to(recv, meth, true)
+      return default if respond == 0
+      unless check_funcall_callable(recv, meth)
+        return check_funcall_missing(recv, meth, args, respond, default);
+      end
+      recv.__send__(meth)
+    end
+
+    def self.check_funcall_respond_to(obj, meth, priv)
+      # TODO Review BJF vm_respond_to
+      return -1 unless object_respond_to_no_built_in?(obj, :respond_to?, true)
+      if !!obj.__send__(:respond_to?, meth, true)
+        1
+      else
+        0
+      end
+    end
+
+    def self.check_funcall_missing(recv, meth, args, respond, default)
+      ret = respond > 0
+      ret = basic_obj_respond_to_missing(recv, meth, false) #PRIV false
+      return default unless ret
+      respond_to_missing = !undefined.equal?(ret)
+      ret = default
+      if object_respond_to_no_built_in?(recv, :method_missing, true)
+        begin
+          return recv.__send__(:method_missing, meth, *args)
+        rescue NoMethodError
+          # TODO BJF usually more is done here
```

Comment:
I think there's some cases that still aren't handled here yet: https://github.com/ruby/ruby/blob/dba03c38c54835248766f28033b5aeddc5f052ef/vm_eval.c#L369